File: src/types/candidate.ts
// ---- Raw backend shapes (schemas/candidate.py: CandidateProfileOut) ----

export interface SkillOut {
    id: string;
    name: string;
    proficiency?: string | null;
    category?: string | null;
    manual_score?: number | null;
    source: string;
  }

  export interface EducationOut {
    id: string;
    degree?: string | null;
    field_of_study?: string | null;
    institution?: string | null;
    start_date?: string | null;
    end_date?: string | null;
    grade?: string | null;
  }

  export interface ExperienceOut {
    id: string;
    title?: string | null;
    company?: string | null;
    industry?: string | null;
    start_date?: string | null;
    end_date?: string | null;
    is_current: boolean;
    description?: string | null;
  }

  export interface CandidateProfileOut {
    id: string;
    user_id: string;
    phone?: string | null;
    location?: string | null;
    summary?: string | null;
    current_designation?: string | null;
    experience_level?: string | null;
    total_experience_years?: number | null;
    experience_years_manual_override: boolean;
    industry?: string | null;
    functional_area?: string | null;
    career_stage?: string | null;
    certifications: string[];
    projects: unknown[];
    achievements: unknown[];
    keywords: string[];
    career_preferences: Record<string, unknown>;
    profile_completeness: number;
    created_at: string;
    updated_at: string;
    skills: SkillOut[];
    education: EducationOut[];
    experiences: ExperienceOut[];
  }

// ---- Input shapes for editing (schemas/candidate.py: *Update / *In) ----

export interface CandidateProfileUpdateInput {
  phone?: string | null;
  location?: string | null;
  summary?: string | null;
  current_designation?: string | null;
  experience_level?: string | null;
  total_experience_years?: number | null;
  industry?: string | null;
  functional_area?: string | null;
  career_stage?: string | null;
}

export interface SkillInput {
  name: string;
  proficiency?: string | null;
  category?: string | null;
}

export interface EducationInput {
  degree?: string | null;
  field_of_study?: string | null;
  institution?: string | null;
  start_date?: string | null;
  end_date?: string | null;
  grade?: string | null;
}

export interface ExperienceInput {
  title?: string | null;
  company?: string | null;
  industry?: string | null;
  start_date?: string | null;
  end_date?: string | null;
  is_current?: boolean;
  description?: string | null;
}

// ---- Recruiter/HR/employer-facing candidate search (schemas/candidate.py: CandidateSearchResultOut) ----

export interface CandidateSearchResult extends CandidateProfileOut {
  full_name: string;
  email: string;
}

export interface CandidateSearchFilters {
  industry?: string;
  functional_area?: string;
  experience_level?: string;
  min_experience_years?: number;
  max_experience_years?: number;
  skill?: string;
  limit?: number;
  offset?: number;
}

export interface CandidateSearchPage {
  items: CandidateSearchResult[];
  total: number;
  limit: number;
  offset: number;
}

File: backend/app/schemas/candidate.py
import uuid
from datetime import date, datetime

from pydantic import BaseModel, ConfigDict, Field


class SkillOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: uuid.UUID
    name: str
    proficiency: str | None = None
    category: str | None = None
    source: str
    manual_score: int | None = None  # user self-rating, 1-5


class SkillIn(BaseModel):
    name: str
    proficiency: str | None = None
    category: str | None = None
    manual_score: int | None = Field(default=None, ge=1, le=5)


class SkillUpdate(BaseModel):
    """Partial update for an existing skill (e.g. set the manual 1-5 score)."""
    proficiency: str | None = None
    category: str | None = None
    manual_score: int | None = Field(default=None, ge=1, le=5)


class EducationIn(BaseModel):
    degree: str | None = None
    field_of_study: str | None = None
    institution: str | None = None
    start_date: date | None = None
    end_date: date | None = None
    grade: str | None = None


class EducationOut(EducationIn):
    model_config = ConfigDict(from_attributes=True)
    id: uuid.UUID


class ExperienceIn(BaseModel):
    title: str | None = None
    company: str | None = None
    industry: str | None = None
    start_date: date | None = None
    end_date: date | None = None
    is_current: bool = False
    description: str | None = None


class ExperienceOut(ExperienceIn):
    model_config = ConfigDict(from_attributes=True)
    id: uuid.UUID


class CareerPreferences(BaseModel):
    desired_roles: list[str] = Field(default_factory=list)
    desired_locations: list[str] = Field(default_factory=list)
    desired_industries: list[str] = Field(default_factory=list)
    min_expected_salary: float | None = None
    remote_preference: str | None = None  # remote/hybrid/onsite


class CandidateProfileUpdate(BaseModel):
    phone: str | None = None
    location: str | None = None
    summary: str | None = None
    current_designation: str | None = None
    experience_level: str | None = None
    total_experience_years: float | None = None
    industry: str | None = None
    functional_area: str | None = None
    career_stage: str | None = None
    career_preferences: CareerPreferences | None = None


class CandidateProfileOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    user_id: uuid.UUID
    phone: str | None
    location: str | None
    summary: str | None
    current_designation: str | None
    experience_level: str | None
    total_experience_years: float | None
    experience_years_manual_override: bool
    industry: str | None
    functional_area: str | None
    career_stage: str | None
    certifications: list
    projects: list
    achievements: list
    keywords: list
    career_preferences: dict
    profile_completeness: float
    created_at: datetime
    updated_at: datetime
    skills: list[SkillOut] = Field(default_factory=list)
    education: list[EducationOut] = Field(default_factory=list)
    experiences: list[ExperienceOut] = Field(default_factory=list)


class CandidateSearchResultOut(CandidateProfileOut):
    """CandidateProfileOut plus the account fields recruiters/HR/employers need
    to identify a candidate (name/email live on the User model, not the profile).
    """

    full_name: str
    # Plain str, not EmailStr — response model echoing a stored address; avoids
    # one bad row 500-ing the directory. See UserOut in schemas/auth.py.
    email: str


class CandidateSearchFilters(BaseModel):
    industry: str | None = None
    functional_area: str | None = None
    experience_level: str | None = None
    min_experience_years: float | None = None
    max_experience_years: float | None = None
    skill: str | None = None


class CandidateSearchPage(BaseModel):
    """Paginated envelope returned by the candidate search endpoint."""
    items: list[CandidateSearchResultOut]
    total: int
    limit: int
    offset: int
